File: analyze_skin_type.py
import sys
import mysql.connector
import json

def fetch_responses(username):
    # Database connection settings
    db_config = {
        'host': 'localhost',
        'user': 'root',
        'password': '',
        'database': 'skinproject'
    }

    # Establish database connection
    try:
        conn = mysql.connector.connect(**db_config)
        cursor = conn.cursor()

        # Fetch form responses from the database based on username
        query = "SELECT * FROM responses WHERE username = %s"
        cursor.execute(query, (username,))
        row = cursor.fetchone()

        if row:
            responses = [
                row[1], row[2], row[3], row[4], row[5],
                row[6], row[7], row[8], row[9], row[10]
            ]
            return responses

    except mysql.connector.Error as err:
        print(f"Error connecting to the database: {err}")

    finally:
        if conn.is_connected():
            cursor.close()
            conn.close()

    return None

def analyze_skin_type(responses_json):
    responses = json.loads(responses_json)
    # Calculate scores for each skin type
    skin_type_scores = {
        "dry": 0,
        "normal": 0,
        "oily": 0,
        "sensitive": 0,
        "combination": 0,
        "acne-prone":0
    }

    # Apply weights based on assigned values
    for idx, answer in enumerate(responses):
        weight = get_weight_for_question(idx + 1, answer)  # Question numbers start from 1
        skin_types = get_associated_skin_types(idx + 1, answer)
        for skin_type in skin_types:
            skin_type_scores[skin_type] += weight

    # Interpret results and provide feedback
    most_likely_type = max(skin_type_scores, key=skin_type_scores.get)
    return most_likely_type


# Get the weight for a given question and answer
def get_weight_for_question(question_number, answer):
    # Implement logic based on assigned weights (high: 2, medium: 1, low: 0.5)
    # Example:
    if question_number == 1 and answer == "a":
        return 2
    elif question_number == 1 and answer == "b":
        return 1
    elif question_number == 1 and answer == "c":
        return 2
    elif question_number == 2 and answer == "a":
        return 1
    elif question_number == 2 and answer == "b":
        return 1
    elif question_number == 2 and answer == "c":
        return 2
    elif question_number == 3 and answer == "a":
        return 2
    elif question_number == 3 and answer == "b":
        return 1
    elif question_number == 3 and answer == "c":
        return 0.5
    elif question_number == 4 and answer == "a":
        return 1
    elif question_number == 4 and answer == "b":
        return 0.5
    elif question_number == 4 and answer == "c":
        return 1
    elif question_number == 5 and answer == "a":
        return 2
    elif question_number == 5 and answer == "b":
        return 1
    elif question_number == 5 and answer == "c":
        return 2
    elif question_number == 6 and answer == "a":
        return 2
    elif question_number == 6 and answer == "b":
        return 1
    elif question_number == 6 and answer == "c":
        return 2
    elif question_number == 7 and answer == "a":
        return 2
    elif question_number == 7 and answer == "b":
        return 1
    elif question_number == 7 and answer == "c":
        return 1
    elif question_number == 8 and answer == "a":
        return 1
    elif question_number == 8 and answer == "b":
        return 0.5
    elif question_number == 8 and answer == "c":
        return 1
    elif question_number == 9 and answer == "a":
        return 2
    elif question_number == 9 and answer == "b":
        return 1
    elif question_number == 9 and answer == "c":
        return 2
    elif question_number == 10 and answer == "a":
        return 2
    elif question_number == 10 and answer == "b":
        return 1
    else: 
        return 2

# Get associated skin types for a question and answer
def get_associated_skin_types(question_number, answer):
    # Implement logic based on decision tree and questionnaire
    # Example:
    if question_number == 1 and answer == "a":
        return ["dry"]
    elif question_number == 1 and answer == "b":
        return ["normal","combination"]
    elif question_number == 1 and answer == "c":
        return ["oily"]
    elif question_number == 2 and answer == "a":
        return ["normal","dry"]
    elif question_number == 2 and answer == "b":
        return ["combination"]
    elif question_number == 2 and answer == "c":
        return ["oily","acne-prone"]
    elif question_number == 3 and answer == "a":
        return ["sensitive"]
    elif question_number == 3 and answer == "b":
        return ["normal"]
    elif question_number == 3 and answer == "c":
        return ["oily"]
    elif question_number == 4 and answer == "a":
        return ["dry"]
    elif question_number == 4 and answer == "b":
        return ["normal","combination"]
    elif question_number == 4 and answer == "c":
        return ["oily","combination"]
    elif question_number == 5 and answer == "a":
        return ["oily","acne-prone"]
    elif question_number == 5 and answer == "b":
        return ["normal","combination"]
    elif question_number == 5 and answer == "c":
        return ["oily"]
    elif question_number == 6 and answer == "a":
        return ["dry"]
    elif question_number == 6 and answer == "b":
        return ["normal","combination"]
    elif question_number == 6 and answer == "c":
        return ["oily"]
    elif question_number == 7 and answer == "a":
        return ["sensitive"]
    elif question_number == 7 and answer == "b":
        return ["normal","combination"]
    elif question_number == 7 and answer == "c":
        return ["oily","acne-prone"]
    elif question_number == 8 and answer == "a":
        return ["dry"]
    elif question_number == 8 and answer == "b":
        return ["normal","combination"]
    elif question_number == 8 and answer == "c":
        return ["oily","combination"]
    elif question_number == 9 and answer == "a":
        return ["dry"]
    elif question_number == 9 and answer == "b":
        return ["normal","combination"]
    elif question_number == 9 and answer == "c":
        return ["oily"]
    elif question_number == 10 and answer == "a":
        return ["dry"]
    elif question_number == 10 and answer == "b":
        return ["normal"]
    else: 
        return ["oily"]

if __name__ == "__main__":
    # Get username from command line arguments
    if len(sys.argv) != 3:
        print("Usage: python analyze_skin_type.py <username>")
        sys.exit(1)
    
    username = sys.argv[2]

    # Fetch responses from the database
    responses = fetch_responses(username)

    if responses:
        responses_json = json.dumps(responses)  # Convert list to JSON string
        result = analyze_skin_type(responses_json)  # Pass JSON string to the function
        print(result,end="")  # Print the skin type analysis result
    else:
        print("No responses found in the database for the user.")

File: analyze_skin_type.php
<?php
// Start the session
session_start();

// Database connection settings
$servername = "localhost";
$username = "root";
$password = "";
$dbname = "skinproject";

// Create connection
$conn = new mysqli($servername, $username, $password, $dbname);

// Check connection
if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

// Fetch form responses from the database based on the logged-in user's username
$username = $_SESSION['username']; // Assuming username is stored in session

// Use prepared statement to prevent SQL injection
$sql = "SELECT * FROM responses WHERE username = ?";
$stmt = $conn->prepare($sql);
$stmt->bind_param("s", $username);
$stmt->execute();
$result = $stmt->get_result();

if ($result->num_rows == 1) {
    // Extract responses from the database
    $row = $result->fetch_assoc();
    $responses = [
        $row['question1'],
        $row['question2'],
        $row['question3'],
        $row['question4'],
        $row['question5'],
        $row['question6'],
        $row['question7'],
        $row['question8'],
        $row['question9'],
        $row['question10']
    ];

    // Encode responses to JSON string
    $responsesJson = json_encode($responses);
    $userInputEscaped = escapeshellarg($responsesJson);

    // Adjust Python script path and command for Windows
    $pythonScriptPath = "C:\\xampp\\htdocs\\project\\analyze_skin_type.py"; // Update this with the correct path
    $command = "python " . escapeshellarg($pythonScriptPath) . " " . $userInputEscaped . " " . $username . " 2>&1"; // Pass username as argument

    // Execute the command
    $output = shell_exec($command);
    // Display the analysis result or error messages
    if ($output !== null) {
        // Update responses table with skin type analysis result
        $updateq_sql = "UPDATE responses SET skintype = ? WHERE username = ?";
        $updateq_stmt = $conn->prepare($updateq_sql);
        $updateq_stmt->bind_param("ss", $output, $username);

        $updatel_sql = "UPDATE login SET skintype = ? WHERE username = ?";
        $updatel_stmt = $conn->prepare($updatel_sql);
        $updatel_stmt->bind_param("ss", $output, $username);

        if ($updateq_stmt->execute() && $updatel_stmt->execute()) {
            $status = 1;
        } else {
            echo "Failed to add skintype (" . $conn->errno . ") " . $conn->error;
        }

        $updateq_stmt->close();
        $updatel_stmt->close();

    } else {
        echo "Error executing Python script or no output received.";
    }
} else {
    echo "No responses found in the database for the logged-in user.";
}

// Close connections
$stmt->close();
$conn->close();
?>

<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Skin Solución</title>
    <style>
        body {
            background-color: rgb(155, 226, 185);
            background-image: url('bgimage23.jpg');
            background-size: cover;
            background-position: center;
            background-attachment: fixed;
            font-family: Helvetica, Arial, Hiragino Sans GB, STXihei, Microsoft YaHei, WenQuanYi Micro Hei, Hind, MS Gothic, Apple SD Gothic Neo, NanumBarunGothic, sans-serif;
            margin: 0;
            padding: 0;
        }

        .container1 {
            max-width: 800px;
            margin: 50px;
            text-align: left;
            font-size: xx-large
        }

        .container2 {
            max-width: 800px;
            margin: 50px auto;
            text-align: center;
            font-size: x-large font-color: #01372d;

        }

        .container {
            max-width: 800px;
            margin: 50px auto;
            text-align: center;
            font-weight: bold;
            font-size: x-large;
        }

        h1 {
            font-size: 2em;
            margin-bottom: 20px;
            color: #01372d;

        }

        h2 {
            font-size: 2em;
            line-height: 1.5;
            color: #01372d;

        }

        a {
            color: #01372d;
            text-decoration: none;

        }
    </style>
</head>

<body>
    <div class="container1">
        <?php echo '<h1>So, ' . $_SESSION['username'] . '!</h1><br>'; ?>
    </div>
    <div class="container2">
        <?php echo '<h2>Your skin type is, ' . $output . ':)</h2><br>'; ?>
    </div>
    <div class="container">

        <a href="http://localhost/project/products.php">
            <p>Let's recommend you products based on your skintype!</p>
        </a>
    </div>

    <?php
    if ($status == 1) {
        $status = "Skintype Successfully Added";
        echo "<script type='text/javascript'>alert('$status');</script>";
        exit();
    }
    ?>
</body>

</html>